File: kotlin/bc-components/src/main/kotlin/com/blockchaincommons/bccomponents/SealedMessage.kt
package com.blockchaincommons.bccomponents

import com.blockchaincommons.bctags.TAG_SEALED_MESSAGE
import com.blockchaincommons.bcur.UR
import com.blockchaincommons.bcur.URCodable
import com.blockchaincommons.dcbor.Cbor
import com.blockchaincommons.dcbor.CborTaggedCodable
import com.blockchaincommons.dcbor.CborTaggedUtils
import com.blockchaincommons.dcbor.Tag
import com.blockchaincommons.dcbor.tagsForValues

/**
 * A sealed message that can only be decrypted by the intended recipient.
 *
 * [SealedMessage] provides a public key encryption mechanism where a message is
 * encrypted with a symmetric key, and that key is then encapsulated using the
 * recipient's public key. This ensures that only the recipient can decrypt the
 * message by first decapsulating the shared secret using their private key.
 *
 * Features:
 * - Anonymous sender: The sender's identity is not revealed in the sealed
 *   message
 * - Authenticated encryption: Message integrity and authenticity are guaranteed
 * - Forward secrecy: Each message uses a different ephemeral key
 * - Post-quantum security options: Can use ML-KEM for quantum-resistant
 *   encryption
 *
 * The structure internally contains:
 * - An [EncryptedMessage] containing the actual encrypted data
 * - An [EncapsulationCiphertext] containing the encapsulated shared secret
 *
 * CDDL:
 * ```
 * SealedMessage = #6.40019([ encrypted_message, encapsulated_key ])
 * ```
 */
class SealedMessage private constructor(
    private val message: EncryptedMessage,
    private val encapsulatedKey: EncapsulationCiphertext,
) : CborTaggedCodable,
    URCodable {

    /** Returns the encrypted message. */
    fun message(): EncryptedMessage = message

    /** Returns the encapsulated key ciphertext. */
    fun encapsulatedKey(): EncapsulationCiphertext = encapsulatedKey

    /**
     * Returns the encapsulation scheme used for this sealed message.
     */
    fun encapsulationScheme(): EncapsulationScheme =
        encapsulatedKey.encapsulationScheme()

    /**
     * Decrypts the message using the recipient's private key.
     *
     * This method performs the following steps:
     * 1. Decapsulates the shared secret using the recipient's private key
     * 2. Uses the shared secret to decrypt the message
     *
     * @param privateKey the private key of the intended recipient
     * @return the decrypted plaintext
     * @throws BcComponentsException.Crypto if decryption fails
     */
    fun decrypt(privateKey: Decrypter): ByteArray {
        val sharedKey = privateKey.decapsulateSharedSecret(encapsulatedKey)
        return sharedKey.decrypt(message)
    }

    // -- equals / hashCode --

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is SealedMessage) return false
        return message == other.message && encapsulatedKey == other.encapsulatedKey
    }

    override fun hashCode(): Int {
        var result = message.hashCode()
        result = 31 * result + encapsulatedKey.hashCode()
        return result
    }

    // -- toString --

    override fun toString(): String =
        "SealedMessage(scheme=${encapsulationScheme()})"

    // -- CBOR --

    override fun cborTags(): List<Tag> = tagsForValues(listOf(TAG_SEALED_MESSAGE))

    override fun untaggedCbor(): Cbor {
        val messageCbor = message.taggedCbor()
        val encapsulatedKeyCbor = encapsulatedKey.taggedCbor()
        return Cbor.fromArray(listOf(messageCbor, encapsulatedKeyCbor))
    }

    // -- Companion --

    companion object {
        /**
         * Creates a new [SealedMessage] by encrypting [plaintext] for the
         * specified [recipient].
         *
         * @param plaintext the message data to encrypt
         * @param recipient the recipient who will be able to decrypt the message
         * @param aad additional authenticated data (optional, not encrypted but
         *   authenticated)
         * @param testNonce optional nonce for deterministic encryption (testing only)
         * @return a new [SealedMessage]
         */
        fun create(
            plaintext: ByteArray,
            recipient: Encrypter,
            aad: ByteArray? = null,
            testNonce: Nonce? = null,
        ): SealedMessage {
            val (sharedKey, encapsulatedKey) = recipient.encapsulateNewSharedSecret()
            val message = sharedKey.encrypt(plaintext, aad, testNonce)
            return SealedMessage(message, encapsulatedKey)
        }

        /** Decodes a [SealedMessage] from untagged CBOR (a two-element array). */
        fun fromUntaggedCbor(cbor: Cbor): SealedMessage {
            val elements = cbor.tryArray()
            if (elements.size != 2) {
                throw BcComponentsException.invalidData(
                    "SealedMessage",
                    "must have exactly 2 elements, got ${elements.size}",
                )
            }
            val message = EncryptedMessage.fromTaggedCbor(elements[0])
            val encapsulatedKey = EncapsulationCiphertext.fromTaggedCbor(elements[1])
            return SealedMessage(message, encapsulatedKey)
        }

        /** Decodes a [SealedMessage] from tagged CBOR. */
        fun fromTaggedCbor(cbor: Cbor): SealedMessage =
            CborTaggedUtils.fromTaggedCbor(
                cbor,
                tagsForValues(listOf(TAG_SEALED_MESSAGE)),
            ) { fromUntaggedCbor(it) }

        /** Decodes a [SealedMessage] from tagged CBOR binary data. */
        fun fromTaggedCborData(data: ByteArray): SealedMessage =
            CborTaggedUtils.fromTaggedCborData(
                data,
                tagsForValues(listOf(TAG_SEALED_MESSAGE)),
            ) { fromUntaggedCbor(it) }

        /** Decodes a [SealedMessage] from a UR. */
        fun fromUr(ur: UR): SealedMessage {
            ur.checkType("crypto-sealed")
            return fromUntaggedCbor(ur.cbor)
        }

        /** Decodes a [SealedMessage] from a UR string. */
        fun fromUrString(urString: String): SealedMessage =
            fromUr(UR.fromUrString(urString))
    }
}

File: csharp/BCComponents/BCComponents/SealedMessage.cs
using BlockchainCommons.BCTags;
using BlockchainCommons.DCbor;

namespace BlockchainCommons.BCComponents;

/// <summary>
/// A sealed message that can only be decrypted by the intended recipient.
/// </summary>
/// <remarks>
/// <para>
/// <see cref="SealedMessage"/> provides a public key encryption mechanism where
/// a message is encrypted with a symmetric key, and that key is then encapsulated
/// using the recipient's public key. This ensures that only the recipient can
/// decrypt the message by first decapsulating the shared secret using their
/// private key.
/// </para>
/// <para>
/// Features:
/// <list type="bullet">
/// <item>Anonymous sender: The sender's identity is not revealed</item>
/// <item>Authenticated encryption: Message integrity and authenticity are guaranteed</item>
/// <item>Forward secrecy: Each message uses a different ephemeral key</item>
/// <item>Post-quantum security options: Can use ML-KEM for quantum-resistant encryption</item>
/// </list>
/// </para>
/// <para>
/// CDDL:
/// <code>
/// SealedMessage = #6.40019([ encrypted_message, encapsulated_key ])
/// </code>
/// </para>
/// </remarks>
public sealed class SealedMessage : IEquatable<SealedMessage>, ICborTaggedEncodable, ICborTaggedDecodable
{
    /// <summary>Gets the encrypted message content.</summary>
    public EncryptedMessage Message { get; }

    /// <summary>Gets the encapsulated key used to encrypt the message.</summary>
    public EncapsulationCiphertext EncapsulatedKey { get; }

    /// <summary>
    /// Creates a new <see cref="SealedMessage"/> from its component parts.
    /// </summary>
    /// <param name="message">The encrypted message.</param>
    /// <param name="encapsulatedKey">The encapsulated key.</param>
    public SealedMessage(EncryptedMessage message, EncapsulationCiphertext encapsulatedKey)
    {
        Message = message;
        EncapsulatedKey = encapsulatedKey;
    }

    /// <summary>
    /// Creates a new <see cref="SealedMessage"/> by encrypting plaintext for the
    /// given recipient.
    /// </summary>
    /// <param name="plaintext">The message data to encrypt.</param>
    /// <param name="recipient">The recipient who will be able to decrypt the message.</param>
    /// <returns>A new <see cref="SealedMessage"/>.</returns>
    public static SealedMessage Create(byte[] plaintext, IEncrypter recipient)
    {
        return CreateWithAad(plaintext, recipient, null);
    }

    /// <summary>
    /// Creates a new <see cref="SealedMessage"/> with additional authenticated data (AAD).
    /// </summary>
    /// <param name="plaintext">The message data to encrypt.</param>
    /// <param name="recipient">The recipient who will be able to decrypt the message.</param>
    /// <param name="aad">Optional additional authenticated data.</param>
    /// <returns>A new <see cref="SealedMessage"/>.</returns>
    public static SealedMessage CreateWithAad(byte[] plaintext, IEncrypter recipient, byte[]? aad)
    {
        return CreateOpt(plaintext, recipient, aad, null);
    }

    /// <summary>
    /// Creates a new <see cref="SealedMessage"/> with options for testing.
    /// </summary>
    /// <param name="plaintext">The message data to encrypt.</param>
    /// <param name="recipient">The recipient who will be able to decrypt the message.</param>
    /// <param name="aad">Optional additional authenticated data.</param>
    /// <param name="testNonce">Optional nonce for deterministic encryption (testing only).</param>
    /// <returns>A new <see cref="SealedMessage"/>.</returns>
    public static SealedMessage CreateOpt(
        byte[] plaintext,
        IEncrypter recipient,
        byte[]? aad,
        Nonce? testNonce)
    {
        var (sharedKey, encapsulatedKey) = recipient.EncapsulateNewSharedSecret();
        var message = sharedKey.Encrypt(plaintext, aad, testNonce);
        return new SealedMessage(message, encapsulatedKey);
    }

    /// <summary>
    /// Decrypts the message using the recipient's private key.
    /// </summary>
    /// <param name="privateKey">The private key of the intended recipient.</param>
    /// <returns>The decrypted plaintext.</returns>
    /// <exception cref="BCComponentsException">
    /// Thrown if the private key does not match, decapsulation fails, or
    /// decryption fails (e.g., message tampering).
    /// </exception>
    public byte[] Decrypt(IDecrypter privateKey)
    {
        var sharedKey = privateKey.DecapsulateSharedSecret(EncapsulatedKey);
        return sharedKey.Decrypt(Message);
    }

    /// <summary>Gets the encapsulation scheme used for this sealed message.</summary>
    public EncapsulationScheme Scheme => EncapsulatedKey.Scheme;

    // --- ICborTagged ---

    /// <summary>Returns the CBOR tags associated with the SealedMessage type.</summary>
    public static IReadOnlyList<Tag> CborTags => GlobalTags.TagsForValues(BcTags.TagSealedMessage);

    // --- ICborTaggedEncodable ---

    /// <summary>
    /// Returns the untagged CBOR representation as a two-element array
    /// [encrypted_message, encapsulated_key].
    /// </summary>
    public Cbor UntaggedCbor()
    {
        var elements = new List<Cbor>
        {
            Message.TaggedCbor(),
            EncapsulatedKey.ToCbor(),
        };
        return Cbor.FromList(elements);
    }

    /// <summary>Returns the tagged CBOR representation.</summary>
    public Cbor TaggedCbor() => Cbor.ToTaggedValue(CborTags[0], UntaggedCbor());

    /// <summary>Returns the tagged CBOR representation.</summary>
    public Cbor ToCbor() => TaggedCbor();

    // --- ICborTaggedDecodable ---

    /// <summary>
    /// Decodes a <see cref="SealedMessage"/> from an untagged CBOR array.
    /// </summary>
    /// <param name="cbor">The untagged CBOR value (must be a two-element array).</param>
    /// <returns>A new <see cref="SealedMessage"/>.</returns>
    public static SealedMessage FromUntaggedCbor(Cbor cbor)
    {
        var elements = cbor.TryIntoArray();
        if (elements.Count != 2)
            throw BCComponentsException.InvalidData("SealedMessage", $"must have two elements, got {elements.Count}");

        var message = EncryptedMessage.FromTaggedCbor(elements[0]);
        var encapsulatedKey = EncapsulationCiphertext.FromCbor(elements[1]);
        return new SealedMessage(message, encapsulatedKey);
    }

    /// <summary>
    /// Decodes a <see cref="SealedMessage"/> from a tagged CBOR value.
    /// </summary>
    /// <param name="cbor">The tagged CBOR value.</param>
    /// <returns>A new <see cref="SealedMessage"/>.</returns>
    public static SealedMessage FromTaggedCbor(Cbor cbor)
    {
        foreach (var tag in CborTags)
        {
            try
            {
                var item = cbor.TryIntoExpectedTaggedValue(tag);
                return FromUntaggedCbor(item);
            }
            catch (CborWrongTagException) { }
            catch (CborWrongTypeException) { }
        }
        throw new CborWrongTypeException();
    }

    // --- IEquatable<SealedMessage> ---

    /// <inheritdoc/>
    public bool Equals(SealedMessage? other)
    {
        if (other is null) return false;
        return Message == other.Message
            && EncapsulatedKey == other.EncapsulatedKey;
    }

    /// <inheritdoc/>
    public override bool Equals(object? obj) => obj is SealedMessage m && Equals(m);

    /// <inheritdoc/>
    public override int GetHashCode() => HashCode.Combine(Message, EncapsulatedKey);

    /// <summary>Tests equality of two SealedMessage instances.</summary>
    public static bool operator ==(SealedMessage? left, SealedMessage? right) =>
        left is null ? right is null : left.Equals(right);

    /// <summary>Tests inequality of two SealedMessage instances.</summary>
    public static bool operator !=(SealedMessage? left, SealedMessage? right) => !(left == right);

    // --- Display ---

    /// <inheritdoc/>
    public override string ToString() =>
        $"SealedMessage(scheme={Scheme}, message={Message})";
}
